pyqtgraph/widgets/MultiAxisPlotWidget.py

```diff
@@ -0,0 +1,429 @@
+            for a in ["left", "bottom", "right", "top"]:
+                plotitem.hideAxis(a)
+            # fix parent legend not showing child charts
+            plotitem.legend = self.pi.legend
+            for k, pos, axis in [["left", [2, 0], y], ["bottom", [3, 1], x], ["right", [2, 2], y], ["top", [1, 1], x]]:
+                # # DO NOT USE, WILL MAKE AXIS UNMATCHED TO DATA
+                # # you can't add the new ones after, it doesn't work for some reason
+                # # hide them instead
+                # plotitem.layout.removeItem(plotitem.layout.itemAt(*pos))
+                if axis.orientation == k:
+                    plotitem.axes[k] = {"item": axis, "pos": pos}
+                # hide plotitem axes
+                plotitem.hideAxis(a)
+        # CHART
+        if chart is None:
+            chart = PlotDataItem(name=name)
+        plotitem.addItem(chart)
+        # keep plotitem inside chart
+        chart.plotItem = plotitem
+        # keeptrack of connections
+        if chart.name not in self._signalConnectionsByChart:
+            self._signalConnectionsByChart[chart.name] = {}
+        # keep axis track
+        chart.axes = [x_axis, y_axis]
+        # keep chart
+        self.charts[name] = chart
+        # create a mapping for this chart and his axis
+        self.axes[x_axis].charts.append(name)
+        self.axes[y_axis].charts.append(name)
+        return chart, plotitem
+
+    def clearLayout(self):
+        """Clears the widget but keeps all created axis and charts."""
+        while self.layout.count() > 0:
+            item = self.layout.itemAt(0)
+            self.layout.removeAt(0)
+            self.scene().removeItem(item)
+        for chart in self.charts.values():
+            self._disconnect_all(chart)
+
+    def makeLayout(self, axes=None, charts=None):
+        """Adds all given axes and charts to the widget.
+
+        Parameters
+        ----------
+        axes : list of str, optional
+            The names associated with the axes to show.
+        charts : list of PlotItems, optional
+            The names associated with the charts to show.
+        """
+        self.clearLayout()
+        _ = self._show_axes(axes)
+        shown_charts = self._show_charts(charts)
+        if len(shown_charts) != 0:
+            top_level = shown_charts[list(shown_charts)[-1]]
+            # FROM "https://github.com/pyqtgraph/pyqtgraph/pull/2010" by herodotus77
+            # propagate mouse actions to charts "hidden" behind
+            for k, c in self.charts.items():
+                self._connect_signals(top_level, c)
+        # MOVE LEGEND TO LAYOUT
+        if self.pi.legend is not None:
+            self.pi.legend.setParentItem(self.pi)
+        self.update()
+
+    def _show_axes(self, axes=None):
+        """Shows all the selected axes."""
+        # SELECT AND ASSEMBLE AXES
+        if axes is None:
+            axes = list(self.axes)
```

Comment:
generally I try and avoiding doing `in` lookups in lists; but given this dict isn't expected to get particularly large (< 10 keys) I think it's fine... commenting mostly for my own reference if when looking at this again to remind myself I already considered keeping things in `dict` form (which would require a modification in `makeLayout`'s call to `_show_axes`)

Reply:
fixed this, also fixed the one in _show_charts